Add SPI for delegating row expression optimizer

tdcmeehan opened the pull request:
## Description
As part of [RFC-0006](https://github.com/prestodb/rfcs/blob/main/RFC-0006-expression-eval.md), we need to support out of process expression evaluation.  Add support for pluggable expression optimization and planner support to utilize the new SPI.

## Motivation and Context
[RFC-0006](https://github.com/prestodb/rfcs/blob/main/RFC-0006-expression-eval.md). See https://github.com/prestodb/presto/pull/24126 for larger changes that include the Presto sidecar as described in the RFC.

## Impact
No impact by default as the old in-memory evaluation is the default.

## Test Plan
Tests have been added.

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes

```
== RELEASE NOTES ==

SPI Changes
* Add ``CoordinatorPlugin#getExpressionOptimizerFactories`` to customize expression evaluation in the Presto coordinator. :pr:`24144`
```

pdabre12 commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -704,6 +706,11 @@ public ShutdownAction getShutdownAction()
 
+    public ExpressionOptimizerManager getExpressionManager()
+    {
+        return injector.getInstance(ExpressionOptimizerManager.class);
```
Can we make this a member variable?

pdabre12 commented on presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java:
```diff
@@ -0,0 +1,105 @@
+    @Inject
+    public ExpressionOptimizerManager(InternalNodeManager nodeManager, FunctionAndTypeManager functionAndTypeManager, NodeInfo nodeInfo)
+    {
+        requireNonNull(nodeManager, "nodeManager is null");
```
Is there a specific need for the `nodeManager` to be of type `InternalNodeManager`. It seems like it gets used for `pluginNodeManager`, why not inject an instance of  `pluginNodeManager` instead?

pdabre12 reviewed (commented):
Thanks @tdcmeehan. 
Just some initial comments

ZacBlanco commented on presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java:
```diff
@@ -512,7 +512,7 @@ public void clusterPoolsMultiCoordinatorCleanup()
     }
 
-    @Test(timeOut = 60_000, groups = {"clusterPoolsMultiCoordinator"})
+    @Test(timeOut = 600_000, groups = {"clusterPoolsMultiCoordinator"})
```
why is this so much higher now?

ZacBlanco commented on presto-tests/src/test/java/com/facebook/presto/memory/TestMemoryManager.java:
```diff
@@ -544,6 +544,7 @@ public void testClusterPoolsMultiCoordinator()
             generalPool = memoryManager.getClusterInfo(GENERAL_POOL);
             reservedPool = memoryManager.getClusterInfo(RESERVED_POOL);
             MILLISECONDS.sleep(10);
+            System.out.println("waiting");
```
remove

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -385,6 +388,7 @@ public TestingPrestoServer(
             eventListenerManager = ((TestingEventListenerManager) injector.getInstance(EventListenerManager.class));
             clusterStateProvider = null;
             planCheckerProviderManager = injector.getInstance(PlanCheckerProviderManager.class);
+            expressionManager.loadExpressionOptimizerFactory();
```
I feel like having an additional method to configure/load the expression optimizer is a bit of an anti-pattern. It creates the potential for bugs if someone is trying to properly create the expression manager by requiring this additional method to be called. I haven't dug much deeper, but is there a reason we can't perform the loading of the factory in the constructor of the `expressionManager`?

tdcmeehan replied:
I believe the reason for this convention is these loading methods might entail something expensive, or that could fail, and it's helpful to do such things outside of Guice's bootstrapping process.  Such failures are often very verbose because Guice returns the error in the context of where it failed during the bootstrapping process, and I think it's cleaner to do it outside of the Guice dependency graph so that the failure is clear and isolated.

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    @Override
+    public ExpressionOptimizer getExpressionOptimizer()
```
A thought I had while looking at the structure of the expression optimizer manager:

```suggestion
    public ExpressionOptimizer getExpressionOptimizer(Session session)
```


Do you think it would be beneficial to add the ability to pass session properties here? I could imagine a case where expression optimization could be configured in the future by the session or potentially switch/update configurations by the end user.

The current implementation requires the user to update config files and restart the server if there were other configs outside of just enabling/disabling the delegating optimizer. It might also allow you to encapsulate the configuration logic for returning the `DelegatingExpressionOptimizer` entirely in this class rather than having the current conditional inside of the `SimplifyRowExpressions` class.

rschlussel replied:
I've been thinking about this for other plugins as well- of having the ability to pass session properties to plugins. Right now we only support session properties for connectors, but we should support them for other plugins too.  When I looked into this before, I think we needed to basically create a connectorSession for these plugins as well and be able to get session properties based on the name prefix of the plugin.  The only thing is that it means you can't e.g. give your catalog and function namespace manager the same name.

tdcmeehan replied:
I think I can get this to work for now by passing in a ConnectorSession, and casting to FullConnectorSession inside of ExpressionOptimizerManager.  I agree this will reduce the branching that both of you have pointed out.  I agree we should have a design around plugin sessions as well.

tdcmeehan replied:
Yes, this will work.  Thanks for the suggestion @ZacBlanco.

It can't be a `Session` because these optimizers are used in connectors as well, and those connectors won't have access to `Session` (see `RowExpressionService`).  But we can pass in a `ConnectorSession`, and convert it back to a `FullConnectorSession` as a compromise.

This eliminates the branch in `SimplifyRowExpressions` and centralizes where one toggles the expression evaluation code.  I think it's cleaner this way in spite of the hack to convert `ConnectorSession` to `FullConnectorSession`.  It also guarantees that you use the right `ExpressionOptimizer` as long as you retrieve it from the appropriate interfaces (`RowExpressionService` in a connector, `ExpressionOptimizerManager` in the planner).  Let me know what you think of this approach.

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/sql/relational/DelegatingRowExpressionOptimizer.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Override
+    public Object optimize(RowExpression rowExpression, Level level, ConnectorSession session, Function<VariableReferenceExpression, Object> variableResolver)
```
not a big fan of having the duplicated code in these methods. Consider refactoring to a private method?

tdcmeehan replied:
It's difficult as it stands because these two methods return different types.  I raised https://github.com/prestodb/presto/pull/24287 which would make it easier to share code between the two methods.

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java:
```diff
@@ -0,0 +1,103 @@
+public class ExpressionOptimizerManager
+        implements ExpressionOptimizerProvider
+{
+    private static final File EXPRESSION_MANAGER_CONFIGURATION = new File("etc/expression-manager.properties");
```
Do we really need a separate config file for this? Is there a reason it can't just be rolled into the existing coordinator properties file?

Additionally, I feel like rather than having a flag which explicitly enables this, then having to configure the specific expression optimizer factory is complex. It would be simpler to have a single flag which specifies the `expression-manager-factory.name` and the default for the flag should be the current default behavior. From what I understand it should default to the name for the in-memory expression optimizer.

tdcmeehan replied:
It's a separate config because the implementation will define properties that the Presto runtime can't know in advance.

I think this should be modeled as a plugin because it introduces a degree of separation between Presto and Velox, and also opens the door for third party integrations.

ZacBlanco reviewed (commented):
At a high-level, the changes make sense. I had a few questions about the configuration and how we might be able to simplify it to be more straightforward for users

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyRowExpressions.java:
```diff
@@ -39,44 +44,52 @@
             RowExpression rewritten = RowExpressionTreeRewriter.rewriteWith(logicalExpressionRewriter, expression, true);
-            RowExpression optimizedRowExpression = optimizer.optimize(rewritten, SERIALIZABLE, session);
-            return optimizedRowExpression;
+            if (isDelegatingRowExpressionOptimizerEnabled(session)) {
```
I think I can do that by following @ZacBlanco's feedback: https://github.com/prestodb/presto/pull/24144#discussion_r1890842187

pdabre12 commented on presto-tests/src/test/java/com/facebook/presto/tests/expressions/TestExpressionOptimizers.java:
```diff
@@ -0,0 +1,133 @@
+                        new ConstantExpression(castVarcharToLikePattern(utf8Slice(pattern)), LIKE_PATTERN)));
+        assertEquals(optimizeRowExpression(predicate, EVALUATED), expected);
+    }
+    @Override
```
Nit: Add an empty line.

pdabre12 commented on presto-main/src/main/java/com/facebook/presto/sql/relational/ConnectorRowExpressionService.java:
```diff
@@ -54,9 +55,9 @@ public DomainTranslator getDomainTranslator()
 
     @Override
-    public ExpressionOptimizer getExpressionOptimizer()
+    public ExpressionOptimizer getExpressionOptimizer(ConnectorSession session)
```
It seems like this change is made in a later commit (https://github.com/prestodb/presto/pull/24144/commits/2d609d0aace7246a922029cf11ca16c72cc1f01b ), whereas the function signature in `RowExpressionService` was changed in prior commit. This will cause the build to fail.
Please make this change in a prior commit.

pdabre12 reviewed (commented):
Some comments,

tdcmeehan commented on presto-tests/src/test/java/com/facebook/presto/tests/expressions/TestExpressionOptimizers.java:
```diff
@@ -0,0 +1,133 @@
+        extends TestExpressions
+{
+    private final FunctionResolution resolution = new FunctionResolution(getMetadata().getFunctionAndTypeManager().getFunctionAndTypeResolver());
+    private ExpressionOptimizer expressionOptimizer;
```
This is constructed in the `setUp` method, so it can't be final.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/sql/relational/DelegatingRowExpressionOptimizer.java:
```diff
@@ -0,0 +1,80 @@
+        for (int i = 0; i < maxOptimizationAttempts; i++) {
+            originalExpression = currentExpression;
+            currentExpression = delegate.optimize(toRowExpression(currentExpression, rowExpression.getType()), level, session, variableResolver);
+            if (currentExpression == null || currentExpression.equals(originalExpression)) {
```
It won't, it's removed in the recent refactoring.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/sql/relational/DelegatingRowExpressionOptimizer.java:
```diff
@@ -0,0 +1,80 @@
+public final class DelegatingRowExpressionOptimizer
+        implements ExpressionOptimizer
+{
+    private static final int DEFAULT_MAX_OPTIMIZATION_ATTEMPTS = 10;
```
There is a session property.  We actually don't need this constant because the session property's default is 10, I'll remove this.

tdcmeehan commented:
Updates:

1. Removed `DelegatingExpressionOptimizer` and all related tests.
2. Added the ability to configure multiple expression optimizers and switch them based on a session property.  Should be useful for fuzzer testing.

Thanks for the comments, please take another @ZacBlanco @rschlussel @aaneja

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/sql/relational/DelegatingRowExpressionOptimizer.java:
```diff
@@ -0,0 +1,80 @@
+        ExpressionOptimizer delegate = expressionOptimizerManager.getExpressionOptimizer(session);
+        Object currentExpression = rowExpression;
+        Object originalExpression;
+        for (int i = 0; i < maxOptimizationAttempts; i++) {
```
This was to guard against a poorly written expression optimizer, but I removed it, as in principle we can leave that responsibility to the plugin.

ZacBlanco commented on presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java:
```diff
@@ -0,0 +1,155 @@
+        checkArgument(!isNullOrEmpty(name), "File name is empty, full path: %s", configurationFile.getAbsolutePath());
+        checkArgument(!name.equals(DEFAULT_OPTIMIZER_KEY), "Cannot name an expression optimizer instance default");
+
+        Map<String, String> properties = new HashMap<>(loadProperties(configurationFile));
+        String factoryName = properties.remove(EXPRESSION_MANAGER_FACTORY_NAME);
+        checkArgument(!isNullOrEmpty(factoryName), "%s does not contain %s", configurationFile, EXPRESSION_MANAGER_FACTORY_NAME);
+        checkArgument(expressionOptimizerFactories.containsKey(factoryName),
+                "ExpressionOptimizerFactory %s is not registered, registered factories: ", factoryName, expressionOptimizerFactories.keySet());
```
This code looks correct, but can we not re-use any of the existing tools from airlift's configuration module to load this? I feel like doing this manually seems more prone to error.

tdcmeehan replied:
We do leverage airlift configuration tools, but within the plugin.  At this layer, we need more flexibility than I think those tools can provide: we need to check if the plugin is the default plugin, and we need to check the factory name to see if it matches with any of the plugins that have been discovered.  These are removed before being sent to the plugin because they're not plugin-specific properties, but once passed to the plugin they will be loaded via airlift as that's the convention established by other plugins (including connectors).

I agree it's a bit brittle with all the `checkArgument`s, but I feel like the unit tests are the best way to ensure this is right, while making sure we don't pass useless properties to the plugins.  Let me know if you can think of a better way to go about it that I didn't think of.

ZacBlanco reviewed (commented):
Two comments, otherwise lgtm

tdcmeehan commented on presto-main/src/test/java/com/facebook/presto/sql/expressions/TestExpressionOptimizerManager.java:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    @Test
+    public void testDefaultMayBeConfigured()
```
Thanks, added a test showing that it's not allowed.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java:
```diff
@@ -0,0 +1,152 @@
+        checkArgument(connectorSession instanceof FullConnectorSession, "connectorSession is not an instance of FullConnectorSession");
+        Session session = ((FullConnectorSession) connectorSession).getSession();
+        String expressionOptimizerInstance = getExpressionOptimizerInstance(session);
+        if (isNullOrEmpty(expressionOptimizerInstance)) {
```
I feel like this logic is unnecessarily fragmented. Can we instead have the session property backed by a config property (like other session properties/configs) and have "default" as the default value for the config? That will also let people set desired configurations for the whole cluster.

tdcmeehan replied:
Good suggestion, done.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -327,6 +327,7 @@ public final class SystemSessionProperties
     public static final String INLINE_PROJECTIONS_ON_VALUES = "inline_projections_on_values";
     public static final String INCLUDE_VALUES_NODE_IN_CONNECTOR_OPTIMIZER = "include_values_node_in_connector_optimizer";
     public static final String SINGLE_NODE_EXECUTION_ENABLED = "single_node_execution_enabled";
+    public static final String EXPRESSION_OPTIMIZER_INSTANCE = "expression_optimizer_instance";
```
maybe call this "expression_optimizer_name"? instance makes it sound like we're providing the object itself.

tdcmeehan replied:
Agreed, done.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java:
```diff
@@ -0,0 +1,152 @@
+
+    public ExpressionOptimizer getExpressionOptimizer(ConnectorSession connectorSession)
+    {
+        checkArgument(connectorSession instanceof FullConnectorSession, "connectorSession is not an instance of FullConnectorSession");
```
i don't like assuming that a session is a connectorSession. it breaks the abstractions.  In this case all we care about is the session property. can we instead add a getSystemProperty() method to ConnectorSession?

tdcmeehan replied:
While I agree this is not ideal, I'm of the opinion that it's worse to expose `getSystemProperty` to the SPI, because this allows all plugins to access system properties that were previously only known to the engine or runtime.  While not ideal, I think this approach is a bit less worse because it's within `presto-main`, and we have examples of this being done for system connector and metadata tables provided to the JDBC driver.  I do think that we should probably think of a better abstraction for how we provide session properties to plugins which are not connectors, and I've added a TODO and will open a Github issue to track it.  Let me know if you think this is reasonable.

rschlussel replied:
Fair criticism and suggestion to punt on resolving this.  i do feel like the problem i'm talking about isn't about plugins that aren't connectors, it's easy enough to implement passing them ConnectorSessions. The thing that makes this require a ConnectorSession vs. a Session is that it gets called from connector optimizers.  and we had similar issue with stats stuff a while back.  The plugin separation break down when connectors call into stuff from presto-main.

pdabre12 commented on presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionOptimizer.java:
```diff
@@ -24,31 +26,38 @@
-        this.metadata = requireNonNull(metadata, "metadata is null");
+        this(requireNonNull(metadata, "metadata is null").getFunctionAndTypeManager());
+    }
+
+    public RowExpressionOptimizer(FunctionMetadataManager functionMetadataManager)
+    {
+        checkArgument(functionMetadataManager instanceof FunctionAndTypeManager, "Expected functionMetadataManager to be instance of FunctionAndTypeManager");
+        this.functionAndTypeManager = (FunctionAndTypeManager) requireNonNull(functionMetadataManager, "functionMetadataManager is null");
     }
```
This check seems unnecessary,

```suggestion

    public RowExpressionOptimizer(FunctionAndTypeManager functionAndTypeManager)
    {
        this.functionAndTypeManager =  requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
    }
```

tdcmeehan replied:
Good catch, done.

pdabre12 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CteProjectionAndPredicatePushDown.java:
```diff
@@ -87,10 +89,14 @@ public class CteProjectionAndPredicatePushDown
+    private final ExpressionOptimizerManager expressionOptimizerManager;
+    private final FeaturesConfig featuresConfig;
 
-    public CteProjectionAndPredicatePushDown(Metadata metadata)
```
`featuresConfig` does not seem to be used, can we remove it?

tdcmeehan replied:
Yes, thanks!

pdabre12 reviewed (commented):
Just two nits, otherwise LGTM.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java:
```diff
@@ -0,0 +1,156 @@
+                properties,
+                new ExpressionOptimizerContext(nodeManager, functionAndTypeManager, functionResolution));
+        expressionOptimizers.put(name, optimizer);
+        if (isDefault) {
```
Originally I was trying to make it so just installing the plugin could easily configure the right default value.  But I'll take that up in a new PR, I've removed the `default` flag.

aditi-pandit reviewed (approved):
+1 for presto-native-execution changes.
